### DISCUSSION TITLE
Developers interested in pulling animated svgs feature?

### DISCUSSION THREAD

**Discussion by williamforster** [fabricjs/fabric.js#9920] [General]
I am on a mission to get animated svgs parsed into fabric objects.

I have made some progress on my fork [https://github.com/williamforster/fabric.js/tree/svgAnimation](), and was wondering if the devs are interested in pulling my code (maybe after a little cleanup).

The code I have written adds a new FabricObject to correspond to the svg <animate> element. This new fabric object (AnimateElement) does not render anything itself, but slots in nicely with the current xml parsing code. At the end of svg parsing, parseSvgDocument.ts then calls the .animate() function on the appropriate Fabric Objects.

I will work on this regardless, but if the devs would like to direct me in a certain direction and pull the results, feel free.

Attached is an example of the sort of thing I would like to get working in fabric.js, and have already mostly got working with exploratory programming.

![animated-svg-example](https://github.com/fabricjs/fabric.js/assets/22677980/7b1da761-b764-4625-81e9-cafda40b1c48)

![down_arrow_floating](https://github.com/fabricjs/fabric.js/assets/22677980/4a806497-449a-405c-b9a1-55a66f0c531a)

link to actual code diff https://github.com/fabricjs/fabric.js/compare/master...williamforster:fabric.js:svgAnimation

link to Animate element specs https://svgwg.org/specs/animations/#AnimateElement

**Comment by asturur** [comment]
reserved

**Comment by asturur** [comment]
reserved

**Comment by asturur** [comment]
With only the code diff view we have to understand the process from a code review, that at this stage is premature.
I see you created a new class, made some changes to the parser etc etc.
The first step would be to be sure to know which are the animate element specs and know which part of the specs we are trying to support.

**Reply by williamforster** [reply]
Sorry I'm not 100% sure what you expect in response. Do you want me to comment [the svg spec](https://svgwg.org/specs/animations/#AnimateElement) and a list of animation features that should be included? Or are you waiting for the other contributors to comment

**Reply by asturur** [reply]
a link to the svg specs was enough, and to be clear in which part of the specs you are trying to support

**Reply by williamforster** [reply]
I was hoping to include the most common features, but not necessarily all.

✅ these are attributes already supported in my fork


AnimateElement:
Animation timing attributes
Begin
✅Dur
✅repeatCount
Fill

Animation value attributes:
✅values
KeyTimes attribute
✅from
✅to
by (numeric only)

AttributeNames:
transform
Transform-origin
✅Position (x,y)
✅Width
✅Height
✅ Radius
Border radius
✅fill Rgba color animation (but not gradients):
✅stroke
✅stroke-width
✅stroke-dashoffset
stroke-opacity would be nice
fill-opacity would be nice
✅animated clipping paths 
Animated masks

✅Font-size


Stretch goals:
AnimateMotion element (moving graphics along a path including rotating heading)
AnimateTransform element
Set element
keySplines attribute
keyPoints attribute with AnimateMotion

Not Supported:
Animated paths
Gradients

**Reply by williamforster** [reply]
I have found the masking algorithm you were talking about over email.
It is from a link [here](https://www.w3.org/TR/css-masking-1/#MaskValues)
<img width="870" alt="Screenshot 2024-06-13 at 3 13 57 PM" src="https://github.com/fabricjs/fabric.js/assets/22677980/7072a8b5-2298-49d1-b142-4c479fa26188">

I will put it here for reference

**Comment by williamforster** [comment]
I have encountered my first difficulty - I cannot match the svg model of groups and transforms to fabric's model of groups. I see I should be looking at the properties originX and originY as a sort of analog to svg's transform-origin.

But I can't understand how originX and originY behave in groups and how to parse an svg with multiple levels of groups and transforms on each one. I have located the[ jsFiddle](https://jsfiddle.net/1ow02gea/244) that is supposed to explain it but that still leaves me scratching my head. For example it appears a FabricObject in a group cannot have a top:-100, as to be rendered outside the group, and be transformed away from the group's originX/originY

Any ideas from the developers as to how this svg feature should be parsed? I guess it has been done for static svgs, so perhaps I should study that code more

**Reply by asturur** [reply]
you have to let the group code regroup objects.
I use the reviver for that and i think i can share some code.
